Refactor test_service_task_cancellation_dag_order and run it against trio

gsalgado opened the pull request:
Closes #12

gsalgado committed:
Refactor test_service_task_cancellation_dag_order and run it against trio

gsalgado committed:
Attempt #1

pipermerriam commented on async_service/trio.py:
```diff
@@ -156,7 +162,19 @@ def cancel(self) -> None:
+            # it spawns any sub-tasks, we know which nursery to run it from.
+            nursery = self._task_nurseries.get(current_task)
+            if nursery is None:
+                nursery = await trio.open_nursery().__aenter__()
```
I think I tried this once and it didn't work out very well.  I'm thinking that *maybe* this will have to be done somewhere like the internal `_run_and_manager_task()` since it lives for the duration of the task and to actually use the context manager API to open the nursery.

gsalgado replied:
Well, this is inside `_run_and_manage_task()`. Or maybe you mean something else?

gsalgado commented:
@pipermerriam I'm struggling a bit with this. I was hoping I'd be able to chain nurseries as new sub-tasks are created and then have trio take care of terminating them in the right order, but it seems nurseries are not made to be used outside of context managers, so although I seem to have managed to chain them it didn't have the expected result. I'm afraid I might have gone down a rabbit hole, so decided to take a break and ask for help. If you have some time to look at this it'd be great, otherwise I'll try again tomorrow and hope that a fresh pair of eyes will help

pipermerriam commented:
I think this is the test that I want to have this functionality pass:

```python
class AsyncResource:
    is_active = False

    def __aenter__(self):
        self.is_active = True
        return self

    def __aexit__(self, ...):
        self.is_active = False

    def check(self):
        if not self.is_active:
            raise Exception("Not active")

class MyService(Service):
    async def run(self):
        async with AsyncResource() as resource:
            assert resource.is_active
            self.manager.run_task(self.use_resource, resource)
            await self.manager.wait_finished()
        assert not resource.is_active

        async def use_resource(self, resource):
            assert resource.is_active
            try:
                await self.manager.wait_finished()
            finally:
                assert resource.is_active
```

Roughly put, the `__aexit__` block should not run until the child task has **fully** finished.

pipermerriam commented:
It would be really nice to have *something* like a hypothesis test for this which generated an arbitrary tree of task dependencies and executed them to be sure that our tests aren't **accidentally** passing simply because of things like the order things get queued in the event loop.  This would probably require injecting some arbitrary `sleep(0)` calls in different places to ensure that it works even in cases where the event loop needs a few passes to finish doing task cleanup.

gsalgado committed:
New test to check tasks terminate in correct order

gsalgado commented:
I've changed the test to be like what you suggested (https://github.com/ethereum/async-service/pull/21/commits/3e42c97ca86eb1eb79b84c985bf9272ad47d2caa) and cleaned up the chained-nurseries approach, but that still doesn't pass the old DAG test because we don't control the order of task cancellation like in the asyncio implementation. Tomorrow I plan to check your approach on #22 and wrap this up one way or another.

pipermerriam commented:
> but that still doesn't pass the old DAG test because we don't control the order of task cancellation

I have no problems with that test being modified to reflect the desired behavior.

pipermerriam commented:
replaced by #31
